Selenium: Ensure URL has a protocol when navigating

osrjv opened the pull request:
Added an override to the `Go To` method/keyword, which autoformats URLs by default. At this point it means appending an `https://` protocol if none is given. This fixes the direct usage, but also indirect through `Open Available Browser`.

It does _not_ fix the usage through plain `Open Browser`, which manipulates the driver object directly. However, due to how DynamicCore and inheritance works, there is no clean way to wrap parent keywords. It would have to be a look-up through the `self.keywords` attribute by string name.

NB: In addition to `http://`, `https://`, `ftp://` type protocols, it should also not prefix when giving browser internal destinations such as `about:config`.

cmin764 commented on packages/main/src/RPA/Browser/Selenium.py:
```diff
@@ -1949,3 +1950,19 @@ def _send_command_and_get_result(self, cmd, params):
+        """
+        if autoformat:
+            url = str(url).strip()
+            if not re.match(r"^\w+:", url):
```
While this indeed works, wouldn't be safer to use `urlsplit/urlparse` from `urllib.parse` like so?
```python
>>> urllib.parse.urlsplit("https://www.google.com/path/to?q=1")
SplitResult(scheme='https', netloc='www.google.com', path='/path/to', query='q=1', fragment='')
>>> urllib.parse.urlsplit("about:config").scheme
'about'
```

cmin764 commented on packages/main/src/RPA/Browser/Selenium.py:
```diff
@@ -1949,3 +1950,19 @@ def _send_command_and_get_result(self, cmd, params):
+            url = str(url).strip()
+            if not re.match(r"^\w+:", url):
+                self.logger.debug("Adding missing protocol to URL (https://)")
+                url = "https://" + url
```
And then you can use `urllib.parse.urlunsplit(...)` over the listed parts above where you just replace the `scheme` with `https`. ([example](https://github.com/robocorp/example-oauth-email/blob/master/OAuth2.py#L16))

cmin764 commented on packages/main/src/RPA/Browser/Selenium.py:
```diff
@@ -1949,3 +1950,19 @@ def _send_command_and_get_result(self, cmd, params):
+                self.logger.debug("Adding missing protocol to URL (https://)")
+                url = "https://" + url
+
+        self.logger.info(f"Opening url '{url}'")
+        self.driver.get(url)
```
Nit:
```suggestion
        super().go_to(url)
```

osrjv replied:
I referred to this in the PR description. This does not work, since there is no direct `go_to` method in the super class. The upstream library uses [DynamicCore](https://github.com/robotframework/PythonLibCore) which does some trickery around how the methods are registered/called.

cmin764 reviewed (approved):
LGTM, no issues observed, but still left a few potential improvements.

Also, why is this _partially_ fixing the [issue](https://github.com/robocorp/rpaframework/issues/501)? Is it because this PR is addressing only the Selenium fix and we remain with the Playwright one? If "yes", would Playwright work just fine with scheme-less URLs? (if "yes", then we can fully close the issue afterwards)

osrjv commented:
> LGTM, no issues observed, but still left a few potential improvements.
> 
> Also, why is this _partially_ fixing the [issue](https://github.com/robocorp/rpaframework/issues/501)? Is it because this PR is addressing only the Selenium fix and we remain with the Playwright one? If "yes", would Playwright work just fine with scheme-less URLs? (if "yes", then we can fully close the issue afterwards)

Partially because the keyword `Open Browser` does not use this method internally, but instead manipulates the Selenium webdriver object directly. We would have to inherit the keyword and then override the `url` argument formatting, but it's a bit annoying because of the reason I mentioned in the other comment (and in OP). Maybe I should still try to make a fix for it.

osrjv commented:
Changed it a bit to take into account review comments.
- Using urllib.parse for split/unsplit
- Added a keyword for setting the default scheme, to allow changing the value for nested calls

cmin764 commented on packages/main/src/RPA/Browser/Selenium.py:
```diff
@@ -54,10 +56,73 @@ def suppress_logging():
         logger.warning, logger.error = logger_warning, logger_error
 
 
+def ensure_scheme(url, default):
```
Type annotations wouldn't hurt here as well.

cmin764 commented on packages/main/src/RPA/Browser/Selenium.py:
```diff
@@ -54,10 +56,73 @@ def suppress_logging():
+        url = ensure_scheme(url, self._default_scheme)
+        super().go_to(url)
+
+    go_to.__doc__ = BrowserManagementKeywords.go_to.__doc__
```
Does its job, but if you feel this is even more elegant, then go for it (not tested):
```python
@keyword
@functools.wraps(BrowserManagementKeywords.go_to)
def go_to(...)
```

osrjv replied:
functools.wraps effectively makes the outer look like the inner, but maybe a bit too much in this inheritance case. it also replaces attributes like `__qualname__` and `__module__`, and I'm not confident that won't have some surprising side-effects.

cmin764 commented on packages/main/src/RPA/Browser/Selenium.py:
```diff
@@ -405,6 +470,11 @@ def __init__(self, *args, **kwargs) -> None:
+
+        # Add inherit/overriden library keywords
+        overrides = [BrowserManagementKeywordsOverride(self)]
+        self.add_library_components(overrides)
```
Perfect! Glad this just works.

cmin764 reviewed (approved):
Definitely this is better, some last nits left, gj!

cmin764 commented:
Also some release notes and `poetry update` (to **main** and **meta** packages) are missing. (even a _minor_ version bump if you plan to release immediately)
